File: AdventOfCode2023/Day2_1.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AdventOfCode2023
{
    public static class Day2_1
    {
        public static void Execute()
        {
            String line;
            try
            {
                StreamReader sr = new StreamReader("../Day2.txt");

                line = sr.ReadLine();

                List<string> games = new List<string>();

                while (line != null)
                {
                    // Console.WriteLine(line);
                    games.Add(line);
                    line = sr.ReadLine();
                }
                sr.Close();

                int numberOfRedCubes = 12;
                int numberOfGreenCubes = 13;
                int numberOfBlueCubes = 14;

                //# go through each line (separated by line)
                //# go through each allCubesDrawn (separated by ;)
                //# go through each cube (separated by ,)
                //#if out of possible range - allCubesDrawn the flag to false
                //#if the flag is still true (aka the game is possible) - add its ID to the sum

                int sumOfIDsOfPossibleGames = 0;

                for (int i = 0; i < games.Count; i++)
                {
                    bool isGamePossible = true;
                    
                    string allCubesDrawn = games[i].Split(":", StringSplitOptions.RemoveEmptyEntries)[1];
                    string[] sets = allCubesDrawn.Split(";", StringSplitOptions.RemoveEmptyEntries);
                    foreach (string set in sets)
                    {
                        string[] cubes = set.Trim().Split(",", StringSplitOptions.RemoveEmptyEntries);
                        foreach (string cube in cubes)
                        {
                            int number = Convert.ToInt32(cube.Trim().Split(" ", StringSplitOptions.RemoveEmptyEntries)[0]);
                            string color = cube.Trim().Split(" ", StringSplitOptions.RemoveEmptyEntries)[1];
                            if (color == "red" &&  number > numberOfRedCubes
                                || color == "green" && number > numberOfGreenCubes
                                || color == "blue" && number > numberOfBlueCubes)
                            {
                                isGamePossible = false;
                            }
                        }
                    }
                    if (isGamePossible)
                    {
                        sumOfIDsOfPossibleGames += i + 1;
                    }
                }
                Console.WriteLine("Sum of IDs of possible games: " + sumOfIDsOfPossibleGames);
            }
            catch (Exception e)
            {
                Console.WriteLine("Exception: " + e.Message);
            }
            finally
            {
                Console.WriteLine("Executing finally block.");
            }
        }
    }
}

File: pythonSolutions/day2_1.py
with open("pythonSolutions\day2.txt") as file:
    lines = file.readlines()

num_of_red_cubes = 12
num_of_green_cubes = 13
num_of_blue_cubes = 14

sum_of_ids_of_possible_games = 0
for line in lines:
    is_game_possible = True

    game, cubes_drawn = line.split(":")
    game_ID = int(game.split(" ")[1])
    sets = cubes_drawn.strip().split(";")

    for s in sets:
        cubes = s.split(",")

        for cube in cubes:  
            cube = cube.strip()
            number, color = cube.split(" ")
            number = int(number)

            if (color == "red" and number > num_of_red_cubes
                or color == "green" and number > num_of_green_cubes
                or color == "blue" and number > num_of_blue_cubes):
                is_game_possible = False

    if is_game_possible:
        sum_of_ids_of_possible_games += game_ID
        
print(sum_of_ids_of_possible_games)